File: DevEnvironment/Base/src/main/java/testbase/core/ResourceReader.java
package testbase.core;

import base.interfaces.ICloseable;
import base.interfaces.IResourceReader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InputStream;
import java.nio.file.Path;

/**
 * The ResourceReader implements a resource reader.
 */
public final class ResourceReader implements IResourceReader, ICloseable {
    private final Path path;
    private final InputStream stream;

    private final Logger log = LoggerFactory.getLogger(this.getClass());

    /**
     * The ResourceReader constructor.
     */
    public ResourceReader(Path path) {
        Conditions.validatePathNotNullOrEmpty(
            path,
            "The path of a resource.");

        this.path = path;
        this.stream = this.open();
    }

    /**
     * Gets a stream of a resource.
     * Use this API when reading binary streams.
     */
    public InputStream getStream() {
        return this.stream;
    }

    /**
     * Reads a string.
     * The stream will reach the end after this method is invoked.
     */
    public String readString() {
        String result = Scanners.scanStream(this.stream);
        return result;
    }

    /**
     * Closes streams.
     */
    public void close() {
        Streams.closeQuietly(this.stream);
    }

    /**
     * Loads a string.
     */
    public static String loadString(Path path) {
        String result;

        try (ResourceReader reader = new ResourceReader(path)) {
            result = reader.readString();
        }

        return result;
    }

    /**
     * Opens a stream.
     */
    private InputStream open() {
        ClassLoader classLoader = getClass().getClassLoader();

        InputStream stream = classLoader.getResourceAsStream(this.path.toString());

        if (stream == null) {
            String errorMessage = "The resource: " + this.path + " failed to get opened as a stream.";

            this.log.error(errorMessage);
            throw new BaseException(errorMessage);
        }

        return stream;
    }
}

File: DevEnvironmentCLion/Base/core/ResourceReader.cpp
#include "PreCompiled.h"
#include "ResourceReader.h"
#include "ResourcePathBuilder.h"

using namespace base;

/**
 * Creates a new resource reader.
 */
IResourceReaderSharedPtr ResourceReader::make(const std::string& path)
{
    return std::make_shared<ResourceReader>(path);
}

/**
 * The ResourceReader constructor.
 */
ResourceReader::ResourceReader(const std::string& path) :
    m_path(path)
{
    open(m_stream, path);
}

/**
 * The ResourceReader destructor.
 */
ResourceReader::~ResourceReader()
{
    m_stream.close();
}

/**
 * Gets a stream of a resource.
 * Use this API when reading binary streams.
 */
std::ifstream& ResourceReader::getStream()
{
    return m_stream;
}

/**
 * Reads a string.
 * The stream will reach the end after this method is invoked.
 */
std::string ResourceReader::readString()
{
    if (!m_stream.is_open())
    {
        std::string errorMessage = "The resource: " + m_path + " is not opened.";
        throw BaseException(errorMessage);
    }

    std::stringstream content;

    std::string currLine;
    while (getline(m_stream, currLine))
    {
        content << currLine << std::endl;
    }

    return content.str();
}

/**
 * Loads a string.
 */
std::string ResourceReader::loadString(const std::string& path)
{
    ResourceReader reader(path);
    return reader.readString();
}

/**
 * Opens a stream.
 */
void ResourceReader::open(std::ifstream& stream, const std::string& path)
{
    std::string resolvedPath = resolvePath(path);
    stream.open(resolvedPath);

    if (!stream.is_open())
    {
        std::string errorMessage = "The resource: " + resolvedPath + " has failed to get opened for reading.";
        throw BaseException(errorMessage);
    }
}

/**
 * Resolves a path of a resource.
 */
std::string ResourceReader::resolvePath(const std::string& path)
{
    std::string basePath;

    base::ResourcePathBuilder builder;
    std::string result = builder
        .addComponent(path)
        .build();

    return result;
}
